### PR TITLE
[inductor][xpu] Re-enable batch_linear_lhs with layout guards

### LINKED ISSUE #193490: Inductor batch_linear_lhs fusion leaks widened row strides to layout-sensitive users (XPU, torch 2.14)
## Summary

On Intel XPU with `torch 2.14`, `torch.compile` auto-enables a pre-grad graph optimization called `batch_linear_lhs` (XPU-only; `config.pre_grad_fusion_options["batch_linear_lhs"] = {"devices": ("xpu",), "min_fuse_set_size": 2}`, enabled by default in 2.14). The pass batches several `Linear(x, W_i)` that share the same input into one wide `mm(x, cat([W_i]).T)` and then `split`s the result back into views.

Those split views inherit the wide GEMM's row stride, so a `Linear` whose output feeds a **layout-sensitive user** — the graph output, a `.is_contiguous()` / `.stride()` query, or an opaque custom op / Triton-kernel wrapper — receives a **non-contiguous tensor with an unexpected stride** instead of the contiguous eager result.

This is a silent-correctness / crash hazard. Many hand-written XPU kernels (including kernels shipped with vLLM) assert their input is contiguous before bespoke pointer arithmetic, so the leaked layout either **crashes** them at the contiguity check or — worse — **silently reads wrong memory** when the check is absent. It currently impacts running Qwen3.5 / Qwen3.8 served through vLLM on XPU with torch 2.14, which is a fairly popular open-source LLM serving path, so the blast radius is real.

## Reproducer

```python
import torch
import torch._inductor.config as inductor_config


@torch.library.custom_op("repro::require_contiguous", mutates_args=())
def require_contiguous(x: torch.Tensor) -> torch.Tensor:
    # Stands in for an XPU kernel that requires a contiguous input.
    if not x.is_contiguous():
        raise RuntimeError(
            f"expected contiguous input, got stride={tuple(x.stride())}, "
            f"shape={tuple(x.shape)}"
        )
    return x.clone()


@require_contiguous.register_fake
def _require_contiguous_fake(x: torch.Tensor) -> torch.Tensor:
    return torch.empty_like(x)


class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.proj_large = torch.nn.Linear(64, 192, bias=False)  # feeds the custom op
        self.proj_a = torch.nn.Linear(64, 16, bias=False)      # pointwise, shared input
        self.proj_b = torch.nn.Linear(64, 16, bias=False)      # pointwise, shared input

    def forward(self, x):
        large = require_contiguous(self.proj_large(x))
        a = torch.sin(self.proj_a(x))
        b = torch.cos(self.proj_b(x))
        return torch.cat((large, a, b), dim=1)


def main():
    print("torch:", torch.__version__)
    print(
        "batch_linear_lhs default config:",
        inductor_config.pre_grad_fusion_options.get("batch_linear_lhs"),
    )
    torch.manual_seed(0)
    module = M().eval()
    x = torch.randn(32, 64)

    with torch.no_grad():
        _eager = module(x)  # eager: proj_large out is contiguous, stride (192, 1)

    opts = {
        "pre_grad_fusion_options": {"batch_linear_lhs": {"min_fuse_set_size": 2}},
        "post_grad_fusion_options": {},
    }
    with torch.no_grad(), inductor_config.patch(**opts):
        try:
            torch.compile(module, fullgraph=True)(x)
            print("RESULT: PASS")
        except Exception as e:
            print("RESULT: FAIL (layout-sensitive op got a non-contiguous view)")
            print(" ", type(e).__name__, str(e))


if __name__ == "__main__":
    main()
```

The `min_fuse_set_size=2` override forces the pass to fire on this small graph so the leak is reproducible; on full models (e.g. vLLM's Qwen) the default threshold triggers it without any override.

### Output (torch 2.14.0.dev20260811+xpu)

```
torch: 2.14.0.dev20260811+xpu
batch_linear_lhs default config: {'devices': ('xpu',), 'min_fuse_set_size': 2}
RESULT: FAIL (layout-sensitive op got a non-contiguous view)
  RuntimeError expected contiguous input, got stride=(224, 1), shape=(32, 192)
```

The fused tensor's row stride is `224` (= `192 + 16`, the two pointwise linears' widths) instead of the expected `192`. The same pattern surfaces in production as a contiguity-check crash inside vLLM XPU kernels feeding Qwen3.5 / Qwen3.8.

## Impact

- **Crash**: kernels that assert `.is_contiguous()` abort at runtime.
- **Silent correctness**: kernels without the assert read memory at the wrong stride.
- Impacts Qwen3.5 / Qwen3.8 served via vLLM on XPU with torch 2.14.

## Notes

- Specific to the `batch_linear_lhs` XPU fusion, auto-enabled by default in 2.14.
- Fix is up in #193295, which makes the pass skip linears whose output reaches a layout-sensitive user.

> The summary, reproducer, and impact statement above were drafted with an AI assistant (Claude Code) and reviewed by me. I reproduced the failure locally on `torch 2.14.0.dev20260811+xpu` with the included script: the layout-sensitive op receives a non-contiguous view with `stride=(224, 1)` instead of the expected contiguous `(192, 1)`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @gujinghui @fengyuan14 @guangyey

### PR DESCRIPTION
**Pull request by xuhancn**
I reviewed the implementation and test results. The AI-assisted summary below accurately describes the problem and intended fix.

> ## Summary
>
> `batch_linear_lhs` was enabled by default on XPU for PyTorch 2.14, then
> temporarily disabled by #194312 after it caused a Qwen3.5/Qwen3.8 regression
> in vLLM. The fusion replaces parallel `Linear(x, W_i)` operations with one
> wide GEMM and split views. Those views inherit the wide GEMM's row stride, so
> they can be non-contiguous even though the original linear outputs were
> contiguous.
>
> Layout-sensitive consumers can therefore observe unexpected strides, fail in
> `view`-like operations, or pass an incompatible layout to an opaque custom
> kernel.
>
> This PR adds a per-linear layout-safety guard and re-enables
> `batch_linear_lhs` by default on XPU. Safe pointwise consumers remain eligible
> for fusion, while unsafe consumers cause only the affected linear to be
> skipped.
>
> Fixes #193490.
> Alternative to #193264.
>
> ## Changes
>
> - Skip fusion when an output reaches the graph output, a layout or storage
>   observer, a mutable user, a non-ATen or higher-order operator, or an operator
>   with an exact-layout constraint.
> - Reject layout-sensitive views such as `view`, `view_as`, `as_strided`,
>   `_unsafe_view`, and `_reshape_alias`.
> - Walk through layout-preserving ATen views and `getitem` nodes from
>   `split`/`chunk`/`unbind`, using schema alias metadata instead of a
>   hand-maintained view list.
> - Stop propagation at ordinary `contiguous()`, while conservatively rejecting
>   layout-sensitive memory formats.
> - Reject unsupported non-1D or unknown bias layouts.
> - Re-enable `batch_linear_lhs` for XPU and add XPU regression coverage for
>   layout observers, view chains, custom operators, mutable consumers, layout
>   constraints, and safe pointwise consumers.
>
> ## Test Plan
>
> ```bash
> source ~/intel/oneapi/setvars.sh
> python test/inductor/test_group_batch_fusion.py -k batch_linear_lhs -v
> conda run -n pytorch_xpu_build_on spin quicklint
> conda run -n pytorch_xpu_build_on lintrunner -a \
>   test/inductor/test_group_batch_fusion.py \
>   torch/_inductor/config.py \
>   torch/_inductor/fx_passes/group_batch_fusion.py
> ```
>
> The focused run completed 23 tests successfully, and lint reported no issues.
>
> Authored with assistance from an AI assistant.

### CODE REVIEW

**Comment by xuhancn** [comment]
Hi @jansel  this PR need to cherry-pick to release branch. Please review it.

**Comment by jansel** [comment]
@claude review these changes

**Comment by xuhancn** [comment]
@claude review these changes

**Comment by xuhancn** [comment]
@claude review these changes

**Review by guangyey — COMMENTED** [review]
Hi @xuhancn  I think the Agent's comment is reasonable. `as_strided` and `view` etc, may crash if fusion happened. WDYT?

**Comment by xuhancn** [comment]
Thanks @guangyey — you're right, and it reproduces exactly as you described: with a 192-wide projection fused against two 16-wide ones, the fused output gets stride (224, 1), and a downstream `view(-1)` fails with `Cannot view a tensor with shape [32, 192] and strides (224, 1) as shape (6144,)`.

Fixed in the latest push (`9ce15d5b5be`):
- `_has_layout_sensitive_user` now treats `view`/`as_strided` consumers as layout-sensitive and walks through the other view-producing aten ops (`reshape`, `permute`, `transpose`, `slice`, `split`, `unsqueeze`, `squeeze`, `expand`, `select`) since the layout propagates to their consumers.
- Handles both call forms: `call_function[target=aten.view]` and `call_method[target="view"]` (e.g. `x.view(-1)`).
- New regression test `test_batch_linear_lhs_skips_view_users` (3 linears, the view-fed one is skipped, the other two still fuse — counter==1).

Combined with your higher-order-op handling, the guard now covers output users, stride queries, custom ops (packet + overload), higher-order wrappers, and view/as_strided.

**Review by guangyey — COMMENTED** [review]
It seems that we could use only one list to replace both `_VIEW_OPS` and `_VIEW_METHODS`

**Comment by xuhancn** [comment]
Good point — consolidated into a single `_VIEW_METHODS` list matched by op name (e.g. `"view"`), and normalized `call_function` targets via `overloadpacket.__name__` so both `aten.view` / `aten.view.default` and the `call_method[target="view"]` form go through the one list. `_VIEW_OPS` is gone. Pushed (`6e7a5fae51f`); SET_LINTER and the batch_linear_lhs tests pass.

File: torch/_inductor/fx_passes/group_batch_fusion.py [hunk 1]
```diff
@@ -600,6 +607,87 @@ def fuse(self, graph: torch.fx.GraphModule, subset: list[torch.fx.Node]):
...
+            "storage_offset",
+            "stride",
+        ):
+            return True
+        if _op_namespace(user.target) not in (None, "aten"):
+            return True
+        kind = _view_op_kind(user)
+        if kind == "crash":
+            return True
+        if kind == "propagate" and _has_layout_sensitive_user(user):
```
Comment (guangyey) [inline torch/_inductor/fx_passes/group_batch_fusion.py]: This recursive call may be a performance bottleneck because the same node may be visited multiple times.

**Comment by xuhancn** [comment]
@claude review these changes

**Comment by xuhancn** [comment]
All round-4 comments landed in `bc7bd814a88`:
- **Recursion** → iterative worklist with a `seen` OrderedSet (each node visited at most once; no RecursionError on deep chains).
- **split/getitem dead end** → `_view_op_kind` now treats `operator.getitem` as layout-preserving, so `split`/`chunk`/`unbind` (tuple → getitem) and plain indexing `lin(x)[0]` are walked through to a downstream view/custom op. Added unit test `test_batch_linear_lhs_split_getitem_walked` (the end-to-end `split()[0].view(-1)` crashes in eager too since the split slice is already non-contiguous, so a graph-level unit test is the right discriminator).
- **match() comment** → rewritten to describe the walk (no longer "direct users only").
- **expensive `getattr(overloadpacket)`** → now dispatches on `isinstance(tgt, OpOverload)` like `_op_namespace`.
- Renamed `_VIEW_METHODS` → `_VIEW_OP_NAMES`; added `detach`/`t`/`chunk`/`unbind`; header comment notes `reshape` may copy (conservative).

Verified on B580: 10 batch_linear_lhs tests pass (incl. the new unit test); lintrunner (SET_LINTER etc.) and flake8/pyfmt clean.

Still open as the pre-cherry-pick gate: confirming the GDN workload still fuses after the two guard widenings (higher-order ops + view walk).

**Comment by xuhancn** [comment]
@claude review these changes

File: torch/_inductor/fx_passes/group_batch_fusion.py [hunk 2]
```diff
@@ -600,6 +607,104 @@ def fuse(self, graph: torch.fx.GraphModule, subset: list[torch.fx.Node]):
...
+        "expand",
+        "select",
+        "split",
+        "chunk",
+        "unbind",
+        "slice",
+        "detach",
+        "t",
+    ]
+)
```
Comment (jansel) [inline torch/_inductor/fx_passes/group_batch_fusion.py]: Is there a cleaner way to do this?  I think we have tags on aten ops that are view ops.

Comment (xuhancn) [reply torch/_inductor/fx_passes/group_batch_fusion.py]: I looked into the tags — here's what I found:

- `torch.Tag.view_copy` only marks the `*_copy` variants (`aten.view_copy`, `aten.reshape_copy`, ...) that functionalization produces; it is **not** present on the aliasing view ops that actually appear in this pre-grad graph (`aten.view`, `detach`, `permute`, `split`, ...), so it can't drive this guard directly.
- The native signal that does identify the aliasing views is the **schema alias annotation**: `Tensor(a) view(...) -> Tensor(a)` (output shares the input's alias set), vs. `sin`/`mm`/`clone` which have no aliasing.

I verified a schema-alias check picks up view/detach/permute/reshape/t/squeeze/unsqueeze/expand/as_strided/select correctly; the tuple-returning views (split/chunk/unbind, `Tensor(a)[]`) and the `contiguous` edge case (aliased but produces a contiguous result — safe, so the walk should stop there) need a little care.

Proposal: replace the hardcoded `_VIEW_OP_NAMES` list with a cached `_is_view_op(op)` built from the schema alias info (lru_cached per op name), keep `_CRASH_VIEW_OPS = {"view", "as_strided"}` (those two genuinely require a compatible layout and can crash on the non-contiguous fused output), and keep the `operator.getitem` propagation for split/indexing. That drops the rot-prone list and stays aligned with functionalization's view semantics.

Want me to make that change?

Comment (xuhancn) [reply torch/_inductor/fx_passes/group_batch_fusion.py]: @claude What do you think?

**Comment by xuhancn** [comment]
@claude review the comments, and give me your suggest: https://github.com/pytorch/pytorch/pull/193295#discussion_r3790588501

File: torch/_inductor/fx_passes/group_batch_fusion.py (hunk 2 above)
Comment (xuhancn) [reply torch/_inductor/fx_passes/group_batch_fusion.py]: Done in `2c921c0f335`. Confirmed the tag doesn't exist on aliasing views (`torch.Tag.view_copy` only marks the `*_copy` variants), so I went with the schema-alias route: `_is_aten_view_name(name)` (lru_cached) uses `OpOverload.is_view` per op name, iterating `packet.overloads()` since `select`/`slice`/`transpose`/`unbind` have no `.default`. The `_CRASH_VIEW_OPS = {view, as_strided, view_as}` set stays hand-written (crash-vs-propagate is op semantics, not alias info). This drops the rot-prone list, covers `view_as`/`narrow`/`unfold`/`.T` etc. for free, and is behavior-preserving on the previously covered set. Verified: 10 batch_linear_lhs tests pass on XPU; lintrunner/flake8/pyfmt clean.

**Comment by xuhancn** [comment]
@claude review these changes

**Review by jansel — CHANGES_REQUESTED** [review]
address automated feedback from claude

**Comment by xuhancn** [comment]
@claude review these changes

**Comment by xuhancn** [comment]
@jansel this is a crash fix — we need to land it and cherry-pick to release/2.14 ASAP. Claude has approved the latest changes, could you please take another look when you get a chance?

File: torch/_inductor/config.py [hunk 3]
```diff
@@ -364,9 +364,11 @@ def prologue_fusion_enabled() -> bool:
 # merge_splits_pass
 # mutate_cat_pass
 # split_cat_pass
+# batch_linear_lhs is disabled until the XPU layout-safety validation lands;
+# it emits non-contiguous split views that break layout-sensitive consumers.
 pre_grad_fusion_options: dict[str, dict[str, Any]] = {
```
Comment (etaf) [inline torch/_inductor/config.py:376]: Suggest:
```
# For each fusion, the "devices" key specifies the devices on which the fusion is enabled.
# eg. pre_grad_fusion_options = {
#     "batch_linear_lhs": {
#         "devices": ("xpu",),
#         "min_fuse_set_size": 2,
#     },
# }
pre_grad_fusion_options: dict[str, dict[str, Any]] = {}
```

Comment (EikanWang) [reply torch/_inductor/config.py:376]: @xuhancn , @etaf's comment makes sense. Let refine it a little bit.

File: torch/_inductor/fx_passes/group_batch_fusion.py (hunk 1 above)
Comment (xuhancn) [reply torch/_inductor/fx_passes/group_batch_fusion.py]: I reviewed the current implementation and agree that this concern is addressed.

> AI-assisted summary: `74df02c19027` replaced the recursive walk with an
> iterative worklist and a `seen` set, so each reachable node is visited at most
> once.

File: torch/_inductor/config.py (hunk 3 above)
Comment (xuhancn) [reply torch/_inductor/config.py:391]: I reviewed the current configuration and the updated scope.

> AI-assisted summary: the comment now documents the `devices` key. The final
> PR intentionally restores the XPU default after the temporary disable in
> #194312, so `pre_grad_fusion_options` is non-empty rather than using the exact
> disabled example originally suggested.

**Comment by xuhancn** [comment]
@claude review these changes
